File: backend/app/crypto.py
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def generate_key() -> bytes:
    """Generate a random 256-bit key."""
    return os.urandom(32)


def encrypt(key: bytes, plaintext: str) -> str:
    """Encrypt plaintext using AES-256-GCM. Returns base64-encoded ciphertext."""
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)
    ciphertext = aesgcm.encrypt(nonce, plaintext.encode(), None)
    combined = nonce + ciphertext
    return base64.urlsafe_b64encode(combined).decode()


def decrypt(key: bytes, encrypted_b64: str) -> str:
    """Decrypt base64-encoded ciphertext."""
    aesgcm = AESGCM(key)
    combined = base64.urlsafe_b64decode(encrypted_b64)
    nonce = combined[:12]
    ciphertext = combined[12:]
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    return plaintext.decode()


def encode_key(key: bytes) -> str:
    """Encode key to base64url string for URL fragment."""
    return base64.urlsafe_b64encode(key).decode().rstrip("=")


def decode_key(key_b64: str) -> bytes:
    """Decode base64url key from URL fragment."""
    padded = key_b64 + "=" * (4 - len(key_b64) % 4)
    return base64.urlsafe_b64decode(padded)

File: frontend/src/lib/crypto.ts
const KEY_LENGTH = 32;
const IV_LENGTH = 12;

export function generateKey(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(KEY_LENGTH));
}

export async function encrypt(
  key: Uint8Array,
  plaintext: string
): Promise<string> {
  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
  const encoded = new TextEncoder().encode(plaintext);

  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key.buffer as ArrayBuffer,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: iv.buffer as ArrayBuffer },
    cryptoKey,
    encoded
  );

  const combined = new Uint8Array(iv.length + ciphertext.byteLength);
  combined.set(iv, 0);
  combined.set(new Uint8Array(ciphertext), iv.length);

  return btoa(String.fromCharCode(...combined))
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

export async function decrypt(
  key: Uint8Array,
  encryptedB64: string
): Promise<string> {
  const padded = encryptedB64.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(padded);
  const bytes = Uint8Array.from(binary, (c) => c.charCodeAt(0));

  const iv = bytes.slice(0, IV_LENGTH);
  const ciphertext = bytes.slice(IV_LENGTH);

  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key.buffer as ArrayBuffer,
    { name: "AES-GCM" },
    false,
    ["decrypt"]
  );

  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: iv.buffer as ArrayBuffer },
    cryptoKey,
    ciphertext
  );

  return new TextDecoder().decode(decrypted);
}

export function encodeKey(key: Uint8Array): string {
  return btoa(String.fromCharCode(...key))
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

export function decodeKey(keyB64: string): Uint8Array {
  const padded = keyB64.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(padded);
  return Uint8Array.from(binary, (c) => c.charCodeAt(0));
}
